src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.EnumerateSplits.cs

```diff
@@ -0,0 +1,270 @@
+        /// </para>
+        /// </remarks>
+        /// <param name="input">The span to search for a match.</param>
+        /// <param name="count">The maximum number of times the split can occur. If 0, all splits are available.</param>
+        /// <param name="startat">The zero-based character position at which to start the search.</param>
+        /// <returns>A <see cref="ValueSplitEnumerator"/> to iterate over the matches.</returns>
+        public ValueSplitEnumerator EnumerateSplits(ReadOnlySpan<char> input, int count, int startat)
+        {
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.count, ExceptionResource.CountTooSmall);
+            }
+
+            if ((uint)startat > (uint)input.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.startat, ExceptionResource.BeginIndexNotNegative);
+            }
+
+            return new ValueSplitEnumerator(this, input, count, startat, RightToLeft);
+        }
+
+        /// <summary>
+        /// Represents an enumerator containing the set of splits around successful matches found by iteratively applying a regular expression pattern to the input span.
+        /// </summary>
+        [StructLayout(LayoutKind.Auto)]
+        public ref struct ValueSplitEnumerator
+        {
+            private readonly Regex _regex;
+            private readonly ReadOnlySpan<char> _input;
+
+            private int _startAt;
+            private (int Index, int Length) _lastMatch;
+            private Range _currentSplit;
+            private int _remainingCount;
+
+            /// <summary>
+            /// Creates an instance of the <see cref="ValueSplitEnumerator"/> for the passed in <paramref name="regex"/> which iterates over <paramref name="input"/>.
+            /// </summary>
+            /// <param name="regex">The <see cref="Regex"/> to use for finding matches.</param>
+            /// <param name="input">The input span to iterate over.</param>
+            /// <param name="count">The maximum number of times the split can occur.</param>
+            /// <param name="startAt">The position from where the engine should start looking for matches.</param>
+            /// <param name="rtl">Whether the engine is matching from right to left.</param>
+            internal ValueSplitEnumerator(Regex regex, ReadOnlySpan<char> input, int count, int startAt, bool rtl)
+            {
+                _regex = regex;
+                _input = input;
+                _startAt = startAt;
+                _lastMatch = (rtl ? input.Length : 0, -1);
+                _remainingCount = count != 0 ? count : int.MaxValue; // Maintain same behavior as Split(..., count: 0, ...), which treats it as effectively infinite.
```

Comment:
should we have a test that produces more than the ~10 results that the current tests max out at? not sure what bug that would expose, unless perhaps some appalling scaling behavior. but it would be easy to have such a test.

Reply:
ok